semantics-3.0/Semantics.hs

```diff
@@ -236,206 +227,219 @@ data ReduceWarning = ReduceNoWarning
+                    Applied newState cont ->
+                        applyAllAux env newState cont rest (warnings ++ warns) (effects ++ effs)
+                    ApplyNoMatchError -> ApplyAllNoMatchError
+    in applyAllAux env state contract inputs [] []
 
 -- PROCESS
 
--- List of signatures needed by a transaction
-type TransactionSignatures = Set Party
-
-data ProcessError = PEReduceError ReduceError
-                  | PEApplyError ApplyError
+data ProcessError = PEAmbiguousSlotIntervalError
+                  | PEApplyNoMatchError
                   | PEIntervalError IntervalError
                   | PEUselessTransaction
-  deriving (Eq,Ord,Show,Read)
+  deriving (Eq,Ord,Show)
 
-type ProcessWarning = ReduceWarning
-type ProcessEffect = ReduceEffect
 
-data ProcessResult = Processed [ProcessWarning]
-                               [ProcessEffect]
-                               TransactionSignatures
+data ProcessResult = Processed [ReduceWarning]
+                               [Payment]
                                TransactionOutcomes
                                State
                                Contract
                    | ProcessError ProcessError
-  deriving (Eq,Ord,Show,Read)
+  deriving (Eq,Ord,Show)
 
 data Transaction = Transaction { txInterval :: SlotInterval
                                , txInputs   :: [Input] }
-  deriving (Eq,Ord,Show,Read)
+  deriving (Eq,Ord,Show)
+
+type TransactionOutcomes = Map Party Money
+
 
+emptyOutcome :: TransactionOutcomes
+emptyOutcome = Map.empty
+
+
+isEmptyOutcome :: TransactionOutcomes -> Bool
+isEmptyOutcome trOut = all (== Lovelace 0) trOut
+
+
+-- Adds a value to the map of outcomes
+addOutcome :: Party -> Money -> TransactionOutcomes -> TransactionOutcomes
+addOutcome party diffValue trOut = let
+    newValue = case Map.lookup party trOut of
+        Just value -> Lovelace (getLovelace value + getLovelace diffValue)
+        Nothing    -> diffValue
+    in Map.insert party newValue trOut
 
--- | Extract necessary signatures from transaction inputs
-getSignatures :: [Input] -> TransactionSignatures
-getSignatures = foldl' addSig Set.empty
-  where
-    addSig acc (IDeposit _ p _)           = Set.insert p acc
-    addSig acc (IChoice (ChoiceId _ p) _) = Set.insert p acc
-    addSig acc INotify                    = acc
 
 
 -- | Extract total outcomes from transaction inputs and outputs
-getOutcomes :: [ReduceEffect] -> [Input] -> TransactionOutcomes
-getOutcomes effect input =
-  foldl' (\acc (p, m) -> addOutcome p m acc) emptyOutcome (incomes ++ outcomes)
-  where
-    incomes = [ (p, m) | ReduceNormalPay p m <- effect ]
-    outcomes = [ (p, m) | IDeposit _ p m <- input ]
+getOutcomes :: [Payment] -> [Input] -> TransactionOutcomes
+getOutcomes effect input = let
+    outcomes = [ (party, money) | Payment party money <- effect ]
+    incomes  = [ (party, money) | IDeposit _ party money <- input ]
```

Comment:
is this some haskell list comprehension magic whereby missing cases are ignored? where are the `IChoice` and `INotify` cases?

Reply:
There are no list comprehensions in PS so I was forced to do something different but actually I think it might be clearer?
```
-- | Extract total outcomes from transaction inputs and outputs
getOutcomes :: List Payment -> List Input -> TransactionOutcomes
getOutcomes effect input = let
    outcomes = map (\(Payment party money) -> Tuple party money) effect
    isDeposit (IDeposit _ party money) = Just (Tuple party money)
    isDeposit _ = Nothing
    incomes = mapMaybe isDeposit input
    in foldl (\acc (Tuple party money) -> addOutcome party money acc)
        mempty
        (outcomes <> incomes)
```

Reply:
> is this some haskell list comprehension magic whereby missing cases are ignored? where are the `IChoice` and `INotify` cases?

It is

Reply:
I scoured the internet and couldn't find mention of this. I think _this_ is the definition of magic 😭